### DISCUSSION TITLE
Coap Communication leads to memory leak

### DISCUSSION THREAD

**Discussion by Pritesh-1011** [openthread/openthread#7866] [Q&A] [answered]
Hi OpenThread community!

I am using the coap communication as the protocol for the data exchange between the devices with the **confirmation** flag enable. So when I  send the data from the Gateway to Endnode, Endnode will send the confirmation to gateway but it leads to some memory leak during the session. And If remove the confirmation flag then there is no memory leak.

Attached the logs file for the Gateway and Endnode for the reference.
Note : Used Openthread version 1.2 
[EndNode.txt](https://github.com/openthread/openthread/files/9043558/EndNode.txt)
[Gateway.txt](https://github.com/openthread/openthread/files/9043559/Gateway.txt)
.

**Comment by jwhui** [comment]
It appears that your child is configured as a sleepy end device. If so, what is the poll period value? See [pollperiod](https://github.com/openthread/openthread/tree/main/src/cli#pollperiod) CLI command.

Note that poll period will determine the latency sending messages (in this case the CoAP ACK message) to the sleepy end device.

**Reply by Pritesh-1011** [reply]
Hello jwhui,

Thank you so much for the reply.
Here the demo code that I am using is for the sleepy end devices. But the devices will only move into sleep mode when I pressed the button otherwise it will work as the active device. 

So when I was facing the issue, my device was not in sleep mode. so Coap ACK messages will be sent instantly.

**Reply by jwhui** [reply]
OK, then what you are observing are the cached responses - see my other answer https://github.com/openthread/openthread/discussions/7866#discussioncomment-3102846.

**Comment by jwhui** [comment, marked as answer]
What you are observing are the cached responses to support message de-duplication - see:
https://github.com/openthread/openthread/blob/b71c1590c1cdbf6707d3e6ff9735b6784e029e64/src/core/coap/coap.hpp#L276
